File: bot/mempool_listener.py
"""
High-speed mempool listener optimized for detecting Bittensor subnet staking
"""

import asyncio
import logging
import time
from typing import Optional, Callable, Dict, Any, List

from substrateinterface import SubstrateInterface
from scalecodec import ScaleBytes
from bot.config import config

logger = logging.getLogger(__name__)


class MempoolListener:

    def __init__(self):

        self.substrates: List[SubstrateInterface] = []
        self.running = False
        self.callbacks: List[Callable] = []

        self.tx_cache: Dict[str, float] = {}

        self.rpc_endpoints = [
            config.subtensor_rpc,
            "wss://entrypoint-finney.opentensor.ai:443",
            "wss://archivelb-finney.opentensor.ai:443",
        ]

    async def connect(self):
        """Connect to multiple substrate nodes"""

        for rpc_url in self.rpc_endpoints:
            try:

                substrate = SubstrateInterface(
                    url=rpc_url,
                    ss58_format=42,
                    use_remote_preset=True
                )

                self.substrates.append(substrate)
                logger.info(f"Connected to {rpc_url}")

                if len(self.substrates) >= 2:
                    break

            except Exception as e:
                logger.warning(f"Failed connecting to {rpc_url}: {e}")

        if not self.substrates:
            raise Exception("No RPC nodes available")

    def register_callback(self, callback: Callable):
        self.callbacks.append(callback)

    async def _fetch_pending_extrinsics(self, substrate: SubstrateInterface):

        try:
            response = await asyncio.to_thread(
                substrate.rpc_request,
                "author_pendingExtrinsics",
                []
            )          
            pending = response.get("result", [])
            return pending

        except Exception as e:
            logger.debug(f"Pending fetch error: {e}")
            return []

    def _decode_extrinsic(self, substrate: SubstrateInterface, extrinsic_hex):

        try:
            extrinsic = substrate.decode_scale(
                "Extrinsic",
                ScaleBytes(extrinsic_hex)
            )

            call = extrinsic["call"]

            call_module = call["call_module"]
            call_function = call["call_function"]

            if call_module != "SubtensorModule":
                return None

            if call_function not in ("add_stake", "add_stake_limit"):
                return None

            args = {
                arg["name"]: arg["value"]
                for arg in call["call_args"]
            }

            netuid = int(args.get("netuid", -1))

            amount = args.get("amount_staked")

            if amount is None:
                return None

            amount = float(amount) / 1e9

            if amount < config.min_wallet_stake:
                return None
            logger.info(f"netuid: {netuid} amount: {amount}")
            return {
                "type": "stake",
                "netuid": netuid,
                "amount": amount,
                "hotkey_ss58": args.get("hotkey"),
                "timestamp": time.time()
            }

        except Exception as e:
            return None

    async def _safe_callback(self, callback, data):

        callback_name = callback.__name__ if hasattr(callback, '__name__') else str(callback)
        
        try:
            await callback(data)
        except Exception as e:
            logger.error(f"Callback failure ({callback_name}): {e}")

    async def _process_extrinsic(self, extrinsic_hex: str, idx: int, total: int, now: float):
        """Process a single extrinsic concurrently"""
        
        tx_hash = str(extrinsic_hex)
        
        # Cache check
        if tx_hash in self.tx_cache:
            if now - self.tx_cache[tx_hash] < 1:
                return

        # Decode extrinsic (run in thread pool since it's CPU-bound)
        decoded = None

        for substrate in self.substrates:
            # Run decode in thread pool to avoid blocking
            decoded = await asyncio.to_thread(
                self._decode_extrinsic,
                substrate,
                extrinsic_hex
            )
            if decoded:
                break

        if not decoded:
            return

        self.tx_cache[tx_hash] = now

        # Execute callbacks
        for callback in self.callbacks:
            asyncio.create_task(
                self._safe_callback(callback, decoded)
            )

    async def _process_mempool(self):

        tasks = [
            self._fetch_pending_extrinsics(sub)
            for sub in self.substrates
        ]

        results = await asyncio.gather(*tasks, return_exceptions=True)

        pending_extrinsics = []

        for r in results:
            if isinstance(r, list):
                pending_extrinsics.extend(r)

        now = time.time()

        # Process all extrinsics concurrently (fire and forget - no need to wait)
        for idx, extrinsic_hex in enumerate(pending_extrinsics):
            asyncio.create_task(
                self._process_extrinsic(extrinsic_hex, idx, len(pending_extrinsics), now)
            )

        

    async def start(self):

        await self.connect()

        self.running = True

        logger.info(
            f"Mempool listener started with {len(self.substrates)} nodes"
        )

        poll_interval = max(0.02, config.mempool_check_interval)

        while self.running:

            try:
                await self._process_mempool()

            except Exception as e:
                logger.error(f"Mempool loop error: {e}")

            sleep_time = poll_interval

            await asyncio.sleep(sleep_time)

    async def stop(self):

        self.running = False

        logger.info("Mempool listener stopped")

File: src/mempool_listener.rs
use anyhow::Result;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tokio::time::{sleep, Duration};
use tracing::{error, info, warn};
use subxt::OnlineClient;
use subxt::config::PolkadotConfig;
use crate::config;

pub type StakeCallback = Arc<dyn Fn(StakeData) + Send + Sync>;

#[derive(Debug, Clone)]
pub struct StakeData {
    pub netuid: i32,
    pub amount: f64,
    pub hotkey_ss58: String,
    pub timestamp: f64,
}

pub struct MempoolListener {
    substrates: Arc<RwLock<Vec<OnlineClient<PolkadotConfig>>>>,
    running: Arc<RwLock<bool>>,
    callbacks: Arc<RwLock<Vec<StakeCallback>>>,
    tx_cache: Arc<RwLock<HashMap<String, f64>>>,
    rpc_endpoints: Vec<String>,
}

impl MempoolListener {
    pub fn new() -> Self {
        let config = config::get_config();
        let mut rpc_endpoints = vec![];
        
        // Get config synchronously for initialization
        let rt = tokio::runtime::Runtime::new().unwrap();
        let subtensor_rpc = rt.block_on(async {
            config.read().await.subtensor_rpc.clone()
        });
        
        rpc_endpoints.push(subtensor_rpc);
        rpc_endpoints.push("wss://entrypoint-finney.opentensor.ai:443".to_string());
        rpc_endpoints.push("wss://archivelb-finney.opentensor.ai:443".to_string());

        Self {
            substrates: Arc::new(RwLock::new(Vec::new())),
            running: Arc::new(RwLock::new(false)),
            callbacks: Arc::new(RwLock::new(Vec::new())),
            tx_cache: Arc::new(RwLock::new(HashMap::new())),
            rpc_endpoints,
        }
    }

    pub async fn connect(&self) -> Result<()> {
        for rpc_url in &self.rpc_endpoints {
            if self.substrates.read().await.len() >= 2 {
                break;
            }

            match OnlineClient::from_url(rpc_url).await {
                Ok(client) => {
                    self.substrates.write().await.push(client);
                    info!("Connected to {}", rpc_url);
                }
                Err(e) => {
                    warn!("Failed connecting to {}: {}", rpc_url, e);
                }
            }
        }

        if self.substrates.read().await.is_empty() {
            anyhow::bail!("No RPC nodes available");
        }

        Ok(())
    }

    pub async fn register_callback(&self, callback: StakeCallback) {
        self.callbacks.write().await.push(callback);
    }

    async fn fetch_pending_extrinsics(
        &self,
        client: &OnlineClient<PolkadotConfig>,
    ) -> Vec<String> {
        // Note: SubXT doesn't have direct author_pendingExtrinsics support
        // We'll need to use RPC calls directly
        // This is a simplified version - you may need to adjust based on actual API
        vec![] // Placeholder - implement actual RPC call
    }

    fn decode_extrinsic(&self, extrinsic_hex: &str) -> Option<StakeData> {
        // Decode extrinsic and extract stake information
        // This is a simplified version - actual implementation would decode SCALE
        // and check for SubtensorModule::add_stake or add_stake_limit calls
        None // Placeholder - implement actual decoding
    }

    async fn process_extrinsic(&self, extrinsic_hex: String, now: f64) {
        let tx_hash = extrinsic_hex.clone();
        
        // Cache check
        {
            let cache = self.tx_cache.read().await;
            if let Some(&cached_time) = cache.get(&tx_hash) {
                if now - cached_time < 1.0 {
                    return;
                }
            }
        }

        // Decode extrinsic
        let decoded = self.decode_extrinsic(&extrinsic_hex);
        
        if let Some(stake_data) = decoded {
            // Update cache
            self.tx_cache.write().await.insert(tx_hash, now);
            
            // Execute callbacks
            let callbacks = self.callbacks.read().await;
            for callback in callbacks.iter() {
                callback(stake_data.clone());
            }
        }
    }

    async fn process_mempool(&self) -> Result<()> {
        let substrates = self.substrates.read().await.clone();
        let mut pending_extrinsics = Vec::new();

        for substrate in &substrates {
            let extrinsics = self.fetch_pending_extrinsics(substrate).await;
            pending_extrinsics.extend(extrinsics);
        }

        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        for extrinsic_hex in pending_extrinsics {
            let listener = self.clone();
            tokio::spawn(async move {
                listener.process_extrinsic(extrinsic_hex, now).await;
            });
        }
        
        Ok(())
    }

    pub async fn start(&self) -> Result<()> {
        self.connect().await?;
        *self.running.write().await = true;

        info!(
            "Mempool listener started with {} nodes",
            self.substrates.read().await.len()
        );

        let config = config::get_config();
        let poll_interval = {
            let config_guard = config.read().await;
            Duration::from_secs_f64(
                config_guard.mempool_check_interval.max(0.02)
            )
        };

        while *self.running.read().await {
            if let Err(e) = self.process_mempool().await {
                error!("Mempool loop error: {}", e);
            }

            sleep(poll_interval).await;
        }

        Ok(())
    }

    pub async fn stop(&self) {
        *self.running.write().await = false;
        info!("Mempool listener stopped");
    }
}

impl Clone for MempoolListener {
    fn clone(&self) -> Self {
        Self {
            substrates: Arc::clone(&self.substrates),
            running: Arc::clone(&self.running),
            callbacks: Arc::clone(&self.callbacks),
            tx_cache: Arc::clone(&self.tx_cache),
            rpc_endpoints: self.rpc_endpoints.clone(),
        }
    }
}
